File: week1_baseline/python/07_the_run_dsl/boukensha/registry.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .errors import UnknownToolError
from .tool import Tool

if TYPE_CHECKING:
    from collections.abc import Callable

    from .context import Context


class Registry:
    def __init__(self, context: Context) -> None:
        self.context = context

    def tool(
        self,
        name: str,
        *,
        description: str,
        parameters: dict[str, Any] | None = None,
        block: Callable[..., Any],
    ) -> Tool:
        tool = Tool(name, description, parameters or {}, block)
        self.context.register_tool(tool)
        return tool

    def dispatch(self, name: str, args: dict[str, Any] | None = None) -> Any:
        tool = self.context.tools.get(name)
        if tool is None:
            raise UnknownToolError(f"No tool registered as '{name}'")
        return tool.block(**(args or {}))

File: week1_baseline/ruby/11_tui/lib/boukensha/registry.rb
require_relative "errors"

module Boukensha
  class Registry
    attr_reader :context

    def initialize(context)
      @context = context
    end

    # All registered Tool structs, by name. Used by Boukensha::MCP::Server
    # to derive tools/list generically, without needing a separate
    # reference to the Context.
    def tools
      @context.tools
    end

    def tool(name, description:, parameters: {}, &block)
      tool = Tool.new(name.to_s, description, parameters, block)
      @context.register_tool(tool)
      tool
    end

    def dispatch(name, args = {})
      tool = @context.tools[name.to_s]
      raise UnknownToolError, "No tool registered as '#{name}'" unless tool
      tool.block.call(**args.transform_keys(&:to_sym))
    end
  end
end